sdk/resource/resource.go

```diff
@@ -21,11 +21,22 @@ import (
 // Resources should be passed and stored as pointers
 // (`*resource.Resource`).  The `nil` value is equivalent to an empty
 // Resource.
+//
+// Note that the Go == operator compares not just the resource attributes but
+// also all other internals of the Resource type. Therefore, Resource values
+// should not be used as map or database keys. In general, the [Resource.Equal]
```

Comment:
Given this assertion ("should not be used as..."), I am a bit confused about the goal of line 38. If the we are saying "don't use Resource as a map key" why are we explicitly requiring it to be usable as a map key on line 38?

I get that formal backward compatibility requires it but if this was never a reasonable thing to do, perhaps we can decide that it is not a practical backward compatibility requirement we are bound by?
Do we have a document that says comparability is a backward compatibility criteria? Or perhaps a document that says "what compiled in the past should compile in the future" as a catch-all guarantee?

Reply:
I would say we do our best to follow https://go.dev/doc/go1compat which is kind of:

> Or perhaps a document that says "what compiled in the past should compile in the future" as a catch-all guarantee?

The Collector does have it documented here: https://github.com/open-telemetry/opentelemetry-collector/blob/main/VERSIONING.md#go-modules

Probably we should also document it in our https://github.com/open-telemetry/opentelemetry-go/blob/main/VERSIONING.md

Reply:
>with some exceptions.

What are the exceptions? The doc you linked to says: "It is intended that programs written to the Go 1 specification will continue to compile and run correctly, unchanged, over the lifetime of that specification". The "continue to compile" phrase doesn't seem to leave much room for exceptions.

Reply:
Please read the doc further :wink: 

https://go.dev/doc/go1compat#expectations

Reply:
I created https://github.com/open-telemetry/opentelemetry-go/issues/6274

Reply:
OK. Can I argue for one more exception? That maintaining accidental comparability of the structs is not something we care about, unless it is explicitly documented (like it is done for attribute.Set)?

In this particular case it seems to create a whole lot of work to keep the Resource comparable, but unclear there is anyone to benefit from this work.

To be clear, I am not advocating for breaking a compatibility promise. I am merely suggesting that what previously was not explicitly promised is added as another exception (since we don't mind having exceptions).

Reply:
No, we need to abide by what we stated when we made the SDK module stable which included compile stability. Softening our definition of stability after-the-fact is not only something that will hurt our reputation, but a direct logical and explicit violation of backwards compatibility.

Reply:
I would rather expect a proof that changing it would not break anyone.

I can imagine one can use the fact that it is comparable in their unit tests. E.g. when checking the output of a custom resource.Detector.

I have never seen a struct which is documented that it is comparable.